wdl-analysis/src/handlers/hover.rs

```diff
@@ -0,0 +1,335 @@
+            continue;
+        };
+        if let Some(content) = find_global_hover_in_doc(imported_doc, token)? {
+            return Ok(Some(content));
+        }
+    }
+
+    Ok(None)
+}
+
+/// Resolves hover information based on the CST of the document.
+///
+/// This inspects the parent [`SyntaxNode`] of the hovered token to
+/// determine the most specific context.
+fn resolve_hover_by_context(
+    parent_node: &SyntaxNode,
+    token: &SyntaxToken,
+    document: &Document,
+    graph: &DocumentGraph,
+) -> Result<Option<String>> {
+    match parent_node.kind() {
+        SyntaxKind::TypeRefNode | SyntaxKind::LiteralStructNode => {
+            if let Some(s) = document.struct_by_name(token.text()) {
+                let root = if let Some(ns_name) = s.namespace() {
+                    // SAFETY: we just found a struct with this namespace name and the document
+                    // guarantees that `document.namespaces` contains a corresponding entry for
+                    // `ns_name`.
+                    let ns = document.namespace(ns_name).unwrap();
+                    let node = graph.get(graph.get_index(ns.source()).unwrap());
+                    node.document().unwrap().root()
+                } else {
+                    document.root()
+                };
+                return Ok(provide_struct_documentation(s, &root));
+            }
+        }
+        SyntaxKind::CallTargetNode => {
+            let target = CallTarget::cast(parent_node.clone()).unwrap();
+            let mut target_names = target.names();
+
+            let (callee_name, ns_name) = match (target_names.next(), target_names.next()) {
+                // Namespaced call
+                (Some(ns), Some(name)) => {
+                    if token.span() == name.span() {
+                        (name, Some(ns))
+                    } else if token.span() == ns.span() {
+                        // namespace identifier hovered
+                        if let Some(ns) = document.namespace(token.text()) {
+                            return Ok(Some(format!(
+                                "```wdl\n(import) {}\n```\nImports from `{}`",
+                                token.text(),
+                                ns.source()
+                            )));
+                        }
+                        return Ok(None);
+                    } else {
+                        return Ok(None);
+                    }
+                }
+                // Local call
+                (Some(name), None) => {
+                    if token.span() == name.span() {
+                        (name, None)
+                    } else {
+                        return Ok(None);
+                    }
+                }
+                _ => return Ok(None),
+            };
+
+            let (target_doc, target_root) = if let Some(ns_name) = ns_name {
+                let Some(ns) = document.namespace(ns_name.text()) else {
```

Comment:
Should this be an `unwrap` instead? It seems like `ns_name` being `Some` would imply that that namespace _must_ exist in the document being searched here. That would eliminate (potentially erroneously) returning `Ok(None)` when we should expect a resolution.